renjin/org/renjin/cran/stringi/stringi.java

```diff
@@ -136,16 +136,38 @@ public static SEXP stri_enc_toutf8(SEXP str, SEXP is_unknown_8bit, SEXP validate
   public static SEXP stri_isempty(SEXP s1) { throw new EvalException("TODO"); }
   public static SEXP stri_join(SEXP s1, SEXP s2, SEXP s3, SEXP s4) { throw new EvalException("TODO"); }
   public static SEXP stri_join_list(SEXP s1, SEXP s2, SEXP s3) { throw new EvalException("TODO"); }
-  public static SEXP stri_join2(SEXP s1, SEXP s2) { throw new EvalException("TODO"); }
+  public static SEXP stri_join2(SEXP s1, SEXP s2) {
+    if (s1.length() <= 0) {
+      return s1;
+    }
+    if (s2.length() <= 0) {
+      return s2;
+    }
+
+    final int length = __max_length(s1, s2);
+    final ListVector e1 = __ensure_length(length, s1);
```

Comment:
Sorry that I didn't catch this before, but I didn't realize what was going on. This `__ensure_length()` construct is a bit problematic for a few reasons.

First, most of these strings arguments are going to be (atomic) character vectors on input, which means they are stored as an array of `String` objects, or potentially in some more compact representation. 

When you call `getElementAsSEXP()` on an `AtomicVector`, the StringVector implementation has to create a *new* SEXP object, along with Attributes and all the rest of the bookkeeping to wrap this individual element. 

Take for an example a case here where `s1` is a StringArrayVector with the single element "r" and the `s2` is a `RowNamesVector` of length 1 million. 

Neither of these StringVectors occupy much memory: the first because there is only one element, and the second, because, regardless of length, the RowNamesVector only stores its length and generates String objects on demand when AtomicVector.getElementAsString() is invoked.

Now, before you've strarted doing any joining, your calls to `__ensure_length` will allocate two `SEXP[1_000_000]` arrays and create 2 million new SEXP objects that are not actually needed.

Instead of calling `__ensure_length`, consider instead simply casting the string arguments to Vector and writing your loop as follows:

    StringVector vector1 = stringi_prepare_arg_string(e1);
    int length1 = e1.length();

    StringVector vector2 = stringi_prepare_arg_string(e2);
    int length2 = e2.length();     
    for(int i=0; i < length; ++i) {
         int index1 = i % length1;
         int index2 = i % length2;
         result[i] = vector1.getElementAsString(index1) + vector2.getElementAsString(string2);
    }

This will be order or magnitudes more efficient for many inputs.

The [stringi_prepare_arg_string](https://github.com/gagolews/stringi/blob/master/src/stri_prepare_arg.cpp#L210) 
method is also important as it handles coercion of the other types of R objects to StringVectors. You can implement this method with simply casting its argument to a StringVector for the moment, and I can help you replicate the rules the other applies at a second stage.

Reply:
Instead of writing the same modulo index over and over, I would prefer to have a delegating vector which would do this modulo indexing for me over the underlying vector.

I wrote five `Recycling<TYPE>Vector` for the atomic vectors, excluding `RawVector`. The code can be reviewed on https://github.com/bedatadriven/renjin/pull/336